src/library/scala/collection/Iterator.scala

```diff
@@ -155,124 +155,115 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
+    private[this] val buffer = {                              // the buffer
+      val k = self.knownSize
+      if (k > 0) new ArrayBuffer[B](k min size)               // if k < size && !partial, buffer will grow on padding
+      else ArrayBuffer.empty[B]
+    }
     private[this] var filled = false                          // whether the buffer is "hot"
-    private[this] var _partial = true                         // whether we deliver short sequences
-    private[this] var pad: Option[() => B] = None             // what to pad short sequences with
+    private[this] var partial = true                          // whether we deliver short sequences
+    private[this] var pad: () => B = null                     // what to pad short sequences with
 
-    /** Public functions which can be used to configure the iterator before use.
+    /** Specifies a fill element used to pad a partial segment
+     *  so that all segments have the same size.
+     *
+     *  If a fill element has not been configured, the last segment
+     *  may be returned with less than `size` elements.
      *
-     *  Pads the last segment if necessary so that all segments will
-     *  have the same size.
+     *  But if the iterator has been configured `withPartial(false)`,
+     *  the partial segment is dropped and any padding element is ignored.
      *
      *  @param x The element that will be appended to the last segment, if necessary.
      *  @return  The same iterator, and ''not'' a new iterator.
      *  @note    This method mutates the iterator it is called on, which can be safely used afterwards.
-     *  @note    This method is mutually exclusive with `withPartial(true)`.
+     *  @note    Padding requires `withPartial(true)`, which is the default setting.
+     *  @group Configuration
      */
     def withPadding(x: => B): this.type = {
-      pad = Some(() => x)
+      pad = () => x
       this
     }
-    /** Public functions which can be used to configure the iterator before use.
+    /** Select whether to drop the last segment if it has less than `size` elements.
+     *
+     *  If this flag is `false`, elements of a partial segment at the end of the iterator
+     *  are not returned.
      *
-     *  Select whether the last segment may be returned with less than `size`
-     *  elements. If not, some elements of the original iterator may not be
-     *  returned at all.
+     *  The flag defaults to `true`.
      *
      *  @param x `true` if partial segments may be returned, `false` otherwise.
      *  @return  The same iterator, and ''not'' a new iterator.
      *  @note    This method mutates the iterator it is called on, which can be safely used afterwards.
      *  @note    This method is mutually exclusive with `withPadding`.
+     *  @group Configuration
      */
     def withPartial(x: Boolean): this.type = {
-      _partial = x
-      if (_partial) // reset pad since otherwise it will take precedence
-        pad = None
```

Comment:
I think we should keep this

Reply:
There's a test to say don't lose the pad function. Since it's "inconceivable" (to quote Princess Bride) that anyone is relying on the old behavior, I made it more sane. (I take the comment to mean, "because of the way I coded it, I have to reset the pad function.") I'm not unwilling to be persuaded either way.

Reply:
The way I see it is that the last setting I provided is the one that matters more. If I said I want padding, then partial or not doesn't matter. If I said I want `partial = true` then it doesn't make sense to pad because that would never return a partial result.

Reply:
Right, I was about to say maybe I inverted the meaning of the settings. Partial means give me my partial segment, otherwise, drop it if no padding, or pad it otherwise.

Reply:
but that is not order dependent. Partial means no pad.

Reply:
I was waiting for a benchmark test to run, but it's taking half the night. I want to be able to benchmark before I fiddle with `PriorityQueue` (just tweaking after touching the scaladoc). I will kill the benchmark. I guess I'll run a benchmark test when I leave for vacation and check the results after I return.

Reply:
Thanks again for your eyes. To recap, my change was to read "withPartial" to mean "keep the partial segment" instead of "return a partial segment".